src/Marten/Events/Daemon/HighWater/HighWaterDetector.cs

```diff
@@ -0,0 +1,169 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Marten.Services;
+using Marten.Storage;
+using Marten.Util;
+using Npgsql;
+
+namespace Marten.Events.Daemon.HighWater
+{
+    internal class HighWaterDetector: IHighWaterDetector
+    {
+        private readonly ITenant _tenant;
+        private readonly NpgsqlCommand _gapDetection;
+        private readonly NpgsqlCommand _stateDetection;
+        private readonly NpgsqlParameter _start;
+        private readonly NpgsqlCommand _updateStatus;
+        private readonly NpgsqlParameter _newSeq;
+        private readonly NpgsqlCommand _findSafeSequence;
+        private readonly NpgsqlParameter _safeTimestamp;
+
+        public HighWaterDetector(ITenant tenant, EventGraph graph)
+        {
+            _tenant = tenant;
+
+            _findSafeSequence = new NpgsqlCommand($@"select min(seq_id) from {graph.DatabaseSchemaName}.mt_events where mt_events.timestamp >= :timestamp");
```

Comment:
We should consider inedxing strategy for those columns (at least as opt-in). For huge number of events those queries may take some time. On the other hand indexes will slow down appending...

Reply:
Also I'm not sure (yet) about all the assumptions (as I didn't go through all of he changes), but we might think about introducing some numeric offset commit instead of using the timestamp. Timestamp for multinode environments might be hit with time skew issues between nodes.

Reply:
The timestamp is assigned by the database, and the only possible usage of it is in this one case to detect stale, missing values in the sequence. We don't key off that otherwise.

And yes to the indexing. Another option is to shard the table by sequence range. I don't see how a materialized view would help with the frequency that this would be getting hit.